### PR TITLE
Issue 17907 improved host listing

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
This is a significant improvement on the 
Service integrations endpoint 
http://localhost:8080/api/v1/service-integrations/{service-key}?per_page=x&page=y&filter=z
Which now includes all the available sites and also supports pagination.
Bear in mind that this pagination is not like the others which simply browse a SQL table using a limit and an offset. This pagination basically takes all the available hosts (From the index for performance purposes) 
and all the hosts configured in the secrets storage repository. And joins them both in memory.
The results are always organized placing the items with an associated configuration on top. (Sorted alphabetically) and then the rest. Meaning all other non-configured hosts. Also ordered alphabetically.

Neither Descendant nor ascendant order can be applied by design.
But you can choose to filter contents by using a **filter** param.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationHelper.java [hunk 1]
```diff
@@ -66,43 +86,65 @@
...
+                        paginationContext.getPerPage(),
+                        paginationContext.getOrderBy(),
+                        orderDirection,
+                        Collections.emptyMap(),
+                        (Function<PaginatedArrayList<SiteView>, ServiceIntegrationView>) paginatedArrayList -> {
+                            final long count = sitesWithConfigurations.size();
+                            return new ServiceIntegrationView(serviceDescriptor, count,
+                                    paginatedArrayList);
+                        });
+        return page;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationHelper.java]: Codacy Issue found: [Consider simply returning the value vs storing it in local variable 'page'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                return new SiteView(host.getIdentifier(), host.getName(), configured);
+            }).collect(Collectors.toList());
+
+            //And then we're done and out of here.
+            final PaginatedArrayList<SiteView> paginatedArrayList = new PaginatedArrayList<>();
+            paginatedArrayList.setTotalResults(totalCount);
+            paginatedArrayList.addAll(siteViews);
+            return paginatedArrayList;
+
+        } catch (Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java [hunk 3]
```diff
@@ -585,11 +607,114 @@ public void Test_Required_Params() throws IOException, DotDataException, DotSecu
...
+        final Map<String, Param> paramMap = ImmutableMap.of(
+                "param1", Param.newParam("val-1", false, Type.STRING, "label", "hint", true),
+                "param2", Param.newParam("val-2", false, Type.STRING, "label", "hint", true),
+                "param3", Param.newParam("val-3", false, Type.STRING, "label", "hint", true)
+        );
+        final long timeMark = System.currentTimeMillis();
+        final List<Host> hosts= new ArrayList<>();
+        final char[] alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();
+        for(char chr :alphabet) {
+           hosts.add(new SiteDataGen().name(chr + "" + timeMark).nextPersisted());
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [Do not add empty strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    private User mockAdminUser() {
+        final User adminUser = mock(User.class);
+        when(adminUser.getUserId()).thenReturn("dotcms.org.1");
+        when(adminUser.getEmailAddress()).thenReturn("admin@dotcms.com");
+        when(adminUser.getFirstName()).thenReturn("Admin");
+        when(adminUser.getLastName()).thenReturn("User");
+        return adminUser;
+    }
+
+    private List<String> mockAllSitesIdentifiers(int allSitesNumber){
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Parameter 'allSitesNumber' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        //System.out.println("Site names: ");
+        for(final String identifier:allSites){
+            final Host host;
+            final String name;
+            if(Host.SYSTEM_HOST.equals(identifier)){
+               name = "System Host";
+               host = mockSite(identifier, name);
+               when(hostAPI.find(eq(identifier),any(User.class), anyBoolean())).thenReturn(host);
+            } else {
+               name = alphabet[i++] + "" +  time;
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Do not add empty strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java [hunk 6]
```diff
@@ -585,11 +607,114 @@ public void Test_Required_Params() throws IOException, DotDataException, DotSecu
...
 
+        final Map<String, Param> paramMap = ImmutableMap.of(
+                "param1", Param.newParam("val-1", false, Type.STRING, "label", "hint", true),
+                "param2", Param.newParam("val-2", false, Type.STRING, "label", "hint", true),
+                "param3", Param.newParam("val-3", false, Type.STRING, "label", "hint", true)
+        );
+        final long timeMark = System.currentTimeMillis();
+        final List<Host> hosts= new ArrayList<>();
+        final char[] alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();
+        for(char chr :alphabet) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [Local variable 'chr' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+ * PaginatorOrdered implementation for objects of type SiteView.
+ * This Pagination isn't very typical in the sense that items are not retrieved from a database.
+ * Sorting and filtering itself happens right here on top of the list of elements itself.
+ */
+public class SiteViewPaginator implements PaginatorOrdered<SiteView> {
+
+    private static final String CONTENT_TYPE_HOST_QUERY = "+contentType:Host +working:true ";
+    private static final String CONTENT_TYPE_HOST_WITH_TITLE_QUERY = "+contentType:Host +working:true +title:*%s*";
+
+    private Supplier<Set<String>> configuredSitesSupplier;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java]: Codacy Issue found: [Private field 'configuredSitesSupplier' could be made final; it is only initialized in the declaration or constructor.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        //Then the rest.
+        final PaginatedArrayList<SiteView> itemsPage2 = paginator
+                .getItems(user, null, 100, limit + 1 , null, null, Collections.emptyMap());
+
+        final List<String> pageNamesPage2 = itemsPage2.stream().map(SiteView::getName).collect(Collectors.toList());
+        Assert.assertTrue(Ordering.<String> natural().isOrdered(pageNamesPage2));
+
+        //Test a page with both mixed configured and non-configured items are sorted.
+        limit = sitesWithIntegrations.size() + 2;
+        final PaginatedArrayList<SiteView> itemsPageMixed = paginator
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Avoid unused local variables such as 'itemsPageMixed'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java [hunk 9]
```diff
@@ -585,11 +607,114 @@ public void Test_Required_Params() throws IOException, DotDataException, DotSecu
...
+                Assert.assertNotNull(hostsBatch1);
+
+                if(currentPage < numberOfPages) {
+                    Assert.assertEquals(pageSize, hostsBatch1.size());
+                } else {
+                    Assert.assertTrue(pageSize <= hostsBatch1.size());
+                }
+
+                List<SiteView> itemsPage;
+                if(currentPage == 1){
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResource.java [hunk 10]
```diff
@@ -111,7 +113,8 @@ public final Response listAvailableServices(@Context final HttpServletRequest re
     public final Response getServiceIntegrationByKey(
             @Context final HttpServletRequest request,
             @Context final HttpServletResponse response,
-            @PathParam("key") final String serviceKey
+            @PathParam("key") final String serviceKey,
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResource.java]: Codacy Issue found: [The String literal "key" appears 4 times in this file; the first occurrence is on line 116](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        when(adminUser.getFirstName()).thenReturn("Admin");
+        when(adminUser.getLastName()).thenReturn("User");
+        return adminUser;
+    }
+
+    private List<String> mockAllSitesIdentifiers(int allSitesNumber){
+        final List<String> allSites = new LinkedList<>();
+        allSites.add(0, Host.SYSTEM_HOST);
+        for(int i=0; i<= allSitesNumber; i++){
+            allSites.add(""+i);
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Do not add empty strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        }
+        final Random random = new Random();
+        final List<String> sitesWithIntegrations = new LinkedList<>();
+        for (int i=0; i <= bound; i++ ){
+            sitesWithIntegrations.add(allSites.get(random.nextInt(bound)));
+        }
+        return new HashSet<>(sitesWithIntegrations);
+    }
+
+    private List<ContentletSearch> mockSearchResults(List<String> allSites){
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Parameter 'allSites' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/view/SiteView.java [hunk 13]
```diff
@@ -25,6 +26,20 @@
...
+        this.secrets = null;
+    }
+
+    /**
+     * If we want to build a secret-less view but showing that the site has integrations.
+     * @param id
+     * @param name
+     * @param configured
+     */
+    public SiteView(final String id, final String name, final boolean configured) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/view/SiteView.java]: Codacy Issue found: [Avoid variables with short names like id](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+            throws DotSecurityException, DotDataException {
+
+        final int max = alphabet.length - 1;
+        final User user = mockAdminUser();
+        final List<String> allSites = mockAllSitesIdentifiers(max);
+        final Set<String> sitesWithIntegrations = mockSitesWithIntegrations(allSites, 10);
+        //System.out.println("Sites with integrations: ");
+        sitesWithIntegrations.forEach(System.out::println);
+        final HostAPI hostAPI = mock(HostAPI.class);
+        long time = System.currentTimeMillis();
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Local variable 'time' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java [hunk 15]
```diff
@@ -134,6 +138,9 @@ public void Test_Create_service_descriptor_Then_Create_Service_Integration_Then_
         );
         final HttpServletRequest request = mock(HttpServletRequest.class);
         final HttpServletResponse response = mock(HttpServletResponse.class);
+
+        when(request.getRequestURI()).thenReturn("/baseURL");
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [The String literal "/baseURL" appears 7 times in this file; the first occurrence is on line 142](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+    public void Test_Get_Items_Sorted_Pages()
+            throws DotSecurityException, DotDataException {
+        final User user = mockAdminUser();
+        final int maxConfigured = 6; //Only the first page is expected to bring back configured items.
+        final List<String> allSites = mockAllSitesIdentifiers(alphabet.length - 1);
+        final HostAPI hostAPI = mock(HostAPI.class);
+        final List<String> allSitesSortedIdentifiers = new LinkedList<>();
+        final long time = System.currentTimeMillis();
+        //final Map<String,String> debugInfo = new TreeMap<>();
+        int i = 0;
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Avoid variables with short names like i](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+        //System.out.println("Site names: ");
+        for(final String identifier:allSites){
+            final Host host;
+            if(Host.SYSTEM_HOST.equals(identifier)){
+                host = mockSite(identifier, "System Host");
+                when(hostAPI.find(eq(identifier),any(User.class), anyBoolean())).thenReturn(host);
+            } else {
+                final String name = alphabet[i++] + "" +  time;
+                System.out.println(name);
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [System.out.println is used](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationHelper.java [hunk 18]
```diff
@@ -27,33 +32,48 @@
...
 
     ServiceIntegrationHelper() {
-        this(APILocator.getServiceIntegrationAPI(), APILocator.getHostAPI());
+        this(APILocator.getServiceIntegrationAPI(), APILocator.getHostAPI(), APILocator.getContentletAPI());
     }
 
-    private static Comparator<ServiceIntegrationView> compareByCountAndName = Comparator
-            .comparing(ServiceIntegrationView::getConfigurationsCount).thenComparing(ServiceIntegrationView::getName);
+    private static Comparator<ServiceIntegrationView> compareByCountAndName = (o1, o2) -> {
+        final int i = Long.compare(o2.getConfigurationsCount(), o1.getConfigurationsCount());
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationHelper.java]: Codacy Issue found: [Avoid variables with short names like i](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+        long time = System.currentTimeMillis();
+        int i = 0;
+        //System.out.println("Site names: ");
+        for(final String identifier:allSites){
+            final Host host;
+            if(Host.SYSTEM_HOST.equals(identifier)){
+                host = mockSite(identifier, "System Host");
+                when(hostAPI.find(eq(identifier),any(User.class), anyBoolean())).thenReturn(host);
+            } else {
+                final String name = alphabet[i++] + "" +  time;
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Do not add empty strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/view/SiteView.java (hunk 13 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/view/SiteView.java]: Codacy Issue found: [Avoid variables with short names like id](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java (hunk 15 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [The String literal "/baseURL" appears 7 times in this file; the first occurrence is on line 142](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+                return new SiteView(host.getIdentifier(), host.getName(), configured);
+            }).collect(Collectors.toList());
+
+            //And then we're done and out of here.
+            final PaginatedArrayList<SiteView> paginatedArrayList = new PaginatedArrayList<>();
+            paginatedArrayList.setTotalResults(totalCount);
+            paginatedArrayList.addAll(siteViews);
+            return paginatedArrayList;
+
+        } catch (Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/SiteViewPaginator.java]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResource.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResource.java]: Codacy Issue found: [The String literal "key" appears 4 times in this file; the first occurrence is on line 116](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+        when(adminUser.getFirstName()).thenReturn("Admin");
+        when(adminUser.getLastName()).thenReturn("User");
+        return adminUser;
+    }
+
+    private List<String> mockAllSitesIdentifiers(final int allSitesNumber){
+        final List<String> allSites = new LinkedList<>();
+        allSites.add(0, Host.SYSTEM_HOST);
+        for(int i=0; i<= allSitesNumber; i++){
+            allSites.add(""+i);
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Do not add empty strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+    public void Test_Get_Items_Sorted_Pages()
+            throws DotSecurityException, DotDataException {
+        final User user = mockAdminUser();
+        final int maxConfigured = 6; //Only the first page is expected to bring back configured items.
+        final List<String> allSites = mockAllSitesIdentifiers(alphabet.length - 1);
+        final HostAPI hostAPI = mock(HostAPI.class);
+        final List<String> allSitesSortedIdentifiers = new LinkedList<>();
+        final long time = System.currentTimeMillis();
+        //final Map<String,String> debugInfo = new TreeMap<>();
+        int i = 0;
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/util/pagination/SiteViewPaginatorUnitTest.java]: Codacy Issue found: [Avoid variables with short names like i](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java [hunk 23]
```diff
@@ -585,11 +607,114 @@ public void Test_Required_Params() throws IOException, DotDataException, DotSecu
...
+                Assert.assertNotNull(hostsBatch1);
+
+                if(currentPage < numberOfPages) {
+                    Assert.assertEquals(pageSize, hostsBatch1.size());
+                } else {
+                    Assert.assertTrue(pageSize <= hostsBatch1.size());
+                }
+
+                List<SiteView> itemsPage;
+                if(currentPage == 1){
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationResourceTest.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)

**Review by jgambarios — APPROVED** [review]

**Pull request merged**